crates/rooch-framework/sources/coin.move

```diff
@@ -0,0 +1,360 @@
+        let coin_info = account_storage::global_borrow_mut<CoinInfo<CoinType>>(ctx, coin_address<CoinType>());
+        coin_info.supply = coin_info.supply - amount;
+        event::emit<BurnEvent>(ctx, BurnEvent {
+            coin_type_info,
+            amount,
+        });
+    }
+
+    /// Burn `coin` from the specified `account` with capability.
+    /// The capability `burn_cap` should be passed as a reference to `BurnCapability<CoinType>`.
+    /// This function shouldn't fail as it's called as part of transaction fee burning.
+    public fun burn_from<CoinType>(
+        ctx: &mut StorageContext,
+        addr: address,
+        amount: u256,
+        burn_cap: &BurnCapability<CoinType>,
+    ) {
+        let coin_store = account_storage::global_borrow_mut<CoinStore<CoinType>>(ctx, addr);
+        let coin_to_burn = extract(&mut coin_store.coin, amount);
+        burn(ctx, coin_to_burn, burn_cap);
+    }
+
+    /// Destroys a zero-value coin. Calls will fail if the `value` in the passed-in `coin` is non-zero
+    /// so it is impossible to "burn" any non-zero amount of `Coin` without having
+    /// a `BurnCapability` for the specific `CoinType`.
+    public fun destroy_zero<CoinType>(zero_coin: Coin<CoinType>) {
+        let Coin { value } = zero_coin;
+        assert!(value == 0, error::invalid_argument(EDestroyOfNonZeroCoin))
+    }
+
+    /// Extracts `amount` from the passed-in `coin`, where the original coin is modified in place.
+    public fun extract<CoinType>(coin: &mut Coin<CoinType>, amount: u256): Coin<CoinType> {
+        assert!(coin.value >= amount, error::invalid_argument(EInSufficientBalance));
+        coin.value = coin.value - amount;
+        Coin { value: amount }
+    }
+
+    /// Extracts the entire amount from the passed-in `coin`, where the original coin is modified in place.
+    public fun extract_all<CoinType>(coin: &mut Coin<CoinType>): Coin<CoinType> {
+        let total_value = coin.value;
+        coin.value = 0;
+        Coin { value: total_value }
+    }
+
+    /// Freeze a CoinStore to prevent transfers
+    public fun freeze_coin_store<CoinType>(
+        ctx: &mut StorageContext,
+        addr: address,
+        _freeze_cap: &FreezeCapability<CoinType>,
+    ) {
+        let coin_store = account_storage::global_borrow_mut<CoinStore<CoinType>>(ctx, addr);
+        coin_store.frozen = true;
+    }
+
+    /// Unfreeze a CoinStore to allow transfers
+    public fun unfreeze_coin_store<CoinType>(
+        ctx: &mut StorageContext,
+        addr: address,
+        _freeze_cap: &FreezeCapability<CoinType>,
+    ) {
+        let coin_store = account_storage::global_borrow_mut<CoinStore<CoinType>>(ctx, addr);
+        coin_store.frozen = false;
+    }
+
+    /// Creates a new Coin with given `CoinType` and returns minting/freezing/burning capabilities.
+    /// The given signer also becomes the account hosting the information about the coin
+    /// (name, supply, etc.).
+    public fun initialize<CoinType>(
```

Comment:
If add `#[private_generic<CoinType>]` to `initialize`, can we remove the `signer` argument?

Reply:
`initialize ` need remain `signer` argument because of `global_move_to` instruction.

add `#[private_generic(CoinType>)` can equals to `coin_address<CoinType>() == addr` assert, but then can't call `initialize ` in coin_entry due to not in current module.
the error stack is `resource type "T0" in function "0x3::coin::initialize" not defined in current module or not allowed`.